File: mobile/lib/core/network/interceptors/country_scope_interceptor.dart
import 'package:dio/dio.dart';
import 'package:vaxiil_mobile/core/country/country_scope_service.dart';

/// Attach X-Timezone / X-Country; hydrate from X-Resolved-Country when empty.
class CountryScopeInterceptor extends Interceptor {
  CountryScopeInterceptor(this._scope);

  final CountryScopeService _scope;

  @override
  void onRequest(RequestOptions options, RequestInterceptorHandler handler) {
    final tz = _scope.timezone;
    if (tz.isNotEmpty) {
      options.headers['X-Timezone'] = tz;
    }
    final iso = _scope.isoCode2;
    if (iso.isNotEmpty) {
      options.headers['X-Country'] = iso;
    }
    handler.next(options);
  }

  @override
  void onResponse(Response<dynamic> response, ResponseInterceptorHandler handler) {
    final resolved = response.headers.value('x-resolved-country') ??
        response.headers.value('X-Resolved-Country');
    _scope.hydrateFromResolvedIso2(resolved);
    handler.next(response);
  }
}

File: web/src/app/core/country/country-scope.interceptor.ts
import { HttpEventType, HttpInterceptorFn } from '@angular/common/http';
import { inject } from '@angular/core';
import { tap } from 'rxjs/operators';

import {
  CountryScopeService,
  RESOLVED_COUNTRY_HEADER,
} from '@/core/country/country-scope.service';

/** Attach X-Timezone / X-Country; hydrate scope from X-Resolved-Country when empty. */
export const countryScopeInterceptor: HttpInterceptorFn = (req, next) => {
  const scope = inject(CountryScopeService);
  if (req.url.includes('/assets/i18n/')) {
    return next(req);
  }

  const headers: Record<string, string> = {
    'X-Timezone': scope.browserTimezone(),
  };
  const iso = scope.isoCode2();
  if (iso) {
    headers['X-Country'] = iso;
  }

  return next(req.clone({ setHeaders: headers })).pipe(
    tap((event) => {
      if (event.type !== HttpEventType.Response) {
        return;
      }
      const resolved = event.headers.get(RESOLVED_COUNTRY_HEADER);
      scope.hydrateFromResolvedHeader(resolved);
    }),
  );
};
